Drivers: i2c: target eeprom extra api

jacmet opened the pull request:
For some use cases it can be interesting to know when the eeprom has been written to by the I2C host (E.G.  once a stop has been received after a write transaction), so add a "changed" callback.

The existing API for accessing the EEPROM data is not ideal, so add getters for the size and data instead and update the sample the use the new APIs

jacmet committed:
i2c: target: eeprom_target: add changed callback

For some use cases it can be interesting to know when the eeprom has been
written to by the I2C host (E.G.  once a stop has been received after a
write transaction), so add a "changed" callback.

Multiple EEPROM instances are supported, so pass the device and user data
as arguments to the callback.

Signed-off-by: Peter Korsgaard <peter@korsgaard.com>

jacmet committed:
i2c: target: eeprom_target: add _get_size()

For some use cases it can be interesting to know how big the EEPROM is, so
add a getter for it.

Signed-off-by: Peter Korsgaard <peter@korsgaard.com>

jacmet committed:
i2c: target: eeprom_target: add {read,write}_data()

The current functions to access the EEPROM data are not ideal, E.G.:

- eeprom_target_program() is hard coded to write to the beginning of the
  EEPROM data

- eeprom_target_read() reads a single byte, so large overhead to read all
  data

So as an alternative, add an interface similar to what is provided by
zephyr/drivers/eeprom.h, E.G.  functions taking an offset and and length.

Signed-off-by: Peter Korsgaard <peter@korsgaard.com>

jacmet committed:
samples: drivers: target_eeprom: drop unregistration

It is not very handy that the EEPROM stops responding after 1 second, so
drop that.

Signed-off-by: Peter Korsgaard <peter@korsgaard.com>

jacmet committed:
samples: drivers: target_eeprom: add a changed callback

And show the usages of the new _get_size() / _read_data() functions.

Signed-off-by: Peter Korsgaard <peter@korsgaard.com>

henrikbrixandersen commented on include/zephyr/drivers/i2c/target/eeprom.h:
```diff
@@ -21,6 +21,23 @@
+ *
+ * @param dev Pointer to the device structure for the driver instance.
+ */
+typedef void (*eeprom_target_changed_handler_t)(const struct device *dev);
```
Please modify this to also take a `void *user_data` argument as described here: https://docs.zephyrproject.org/latest/develop/api/design_guidelines.html#using-callbacks

jacmet replied:
So in addition to the dev argument? Sure, will change

henrikbrixandersen commented on drivers/i2c/target/eeprom_target.c:
```diff
@@ -50,6 +50,13 @@ unsigned int eeprom_target_get_size(const struct device *dev)
+{
+	struct i2c_eeprom_target_data *data = dev->data;
+
+	return data->buffer;
```
This is typically frowned upon as it breaks encapsulation. You are exposing driver internal data structures here. How about adding proper read/write APIs (similar to those in `include/zephyr/drivers/eeprom.h`) and deprecate the old APIs?

jacmet replied:
That is also fine by me. How exactly do I go about deprecating the old API? Any suggestion for the function names given that the legacy API already has a `eeprom_target_read()`?

henrikbrixandersen reviewed (changes requested):
Thanks! A few comments below:

kartben commented on include/zephyr/drivers/i2c/target/eeprom.h:
```diff
@@ -21,6 +21,41 @@
+ *
+ * @param dev Pointer to the device structure for the driver instance.
+ *
+ * @retval Size of EEPROM in bytes
```
```suggestion
 * @return Size of EEPROM in bytes
```

jacmet replied:
Ahh, thanks!

kartben reviewed (dismissed):
A couple minor doxygen mistakes. thanks!

jacmet committed:
i2c: target: target_eeprom: mark eeprom_target_program as deprecated

It does not provide an offset argument, so suggest
eeprom_target_write_data() instead.

Signed-off-by: Peter Korsgaard <peter@korsgaard.com>

jacmet committed:
tests: drivers: i2c: i2c_target_api: switch to eeprom_target_write_data()

eeprom_target_program() is about to become deprecated as it does not take
an offset argument, so switch to eeprom_target_write_data() instead.

Signed-off-by: Peter Korsgaard <peter@korsgaard.com>

jacmet commented:
@henrikbrixandersen Nerry Christmas! What is still needed to get this merged?

henrikbrixandersen commented on include/zephyr/drivers/i2c/target/eeprom.h:
```diff
@@ -82,15 +82,17 @@ int eeprom_target_write_data(const struct device *dev, off_t offset,
  */
-int eeprom_target_program(const struct device *dev, const uint8_t *eeprom_data,
-			 unsigned int length);
+__deprecated int eeprom_target_program(const struct device *dev, const uint8_t *eeprom_data,
```
Can this deprecated API call be replaced with an inline call to the new write data API?

henrikbrixandersen reviewed (changes requested):
> @henrikbrixandersen Nerry Christmas! What is still needed to get this merged?

Happy New Year! And my apologies for not getting back to you earlier on this PR; it somehow managed to escape my list of TODOs.

The proposed changes look very good, thanks, just a few minor comments below:

henrikbrixandersen commented:
Oh, and the deprecation must be mentioned in the migration guide for v4.4.

jacmet commented on include/zephyr/drivers/i2c/target/eeprom.h:
I didn't deprecate it as it works, just a bit inconvenient for "large" amounts of data, but E.G. it is handy for the hexdump I do in the example.

I can make it call the new API though

henrikbrixandersen replied:
Okay, let's leave it in as a utility wrapper using the new API.

jacmet commented:
> Oh, and the deprecation must be mentioned in the migration guide for v4.4.

OK, I will add a note about it.

jacmet committed:
i2c: target: target_eeprom: make legacy functions wrappers

Implement eeprom_target_program() using eeprom_target_write_data() and
eeprom_target_read() using eeprom_target_read_data() to limit redundant
code.

Signed-off-by: Peter Korsgaard <peter@korsgaard.com>

jacmet commented:
@henrikbrixandersen changes since v5:
- Only read callback once in `eeprom_target_stop()` to protect against TOCTOU issues
- Implement legacy functions as inline wrappers for the new ones
- Add note about the API change in the migration guide
- Use `eeprom_target_read_data()` in sample for consistency

jacmet commented:
Changes since v6:
- Reworded "i2c: target: target_eeprom: make legacy functions wrappers for eeprom_target_{read,write}_data()" title to make CI happy, no functional change

teburd reviewed (approved):
Seems like a nice set of changes
